### PR TITLE
Backend/feat: #1460/driver-profile-summary

### PR DESCRIPTION
**Pull request by prashant601**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
created an API for driver profile summary with driver stats, also  backfilling the stats of driver

### Additional Changes

- [x] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/DriverInformation.hs [hunk 1]
```diff
@@ -75,6 +76,34 @@ data DriverInformationE e = DriverInformation
...
+    bonusEarned :: Money,
+    totalCompletedTrips :: Int,
+    lateNightTrips :: Int,
+    lastRegistered :: UTCTime
+  }
+  deriving (Generic, ToJSON, FromJSON, ToSchema, Show)
+
+data DriverMissedOpp = DriverMissedOpp
+  { cancellationRate :: Int,
+    ridesCancelled :: Maybe Int,
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/DriverInformation.hs:90]: Why maybe?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs [hunk 2]
```diff
@@ -91,18 +81,70 @@ eventPayloadHandler DST.OnDriverCancellation {..} = do
     nonZero (Just a)
       | a <= 0 = 1
       | otherwise = a
+eventPayloadHandler DST.OnRideCompletion {..} = do
+  mbDriverStats <- Esq.runInReplica $ DSQ.findById (cast driverId) -- always be just because stats will be created at OnNewRideAssigned
+  case mbDriverStats of
+    Just driverStats -> do
+      allRides <- if driverStats.totalEarnings == 0 && driverStats.bonusEarned == 0 && driverStats.lateNightTrips == 0 && driverStats.earningsMissed == 0 then Esq.runInReplica $ RQ.findAllRidesByDriverId driverId else pure []
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs:88]: do you mean `unless`?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs [hunk 3]
```diff
@@ -91,18 +81,70 @@ eventPayloadHandler DST.OnDriverCancellation {..} = do
...
+          then do
+            Esq.runInReplica $ FPQ.findAllLateNightRides farePramIds
+          else isLateNightRide ride
+
+      Esq.runNoTransaction $ do
+        DSQ.incrementTotalEarningsAndBonusEarned (cast driverId) incrementTotaEarningsBy (incrementBonusEarningsBy + Money (length farePramIds * 10))
+        DSQ.incrementLateNightTrips (cast driverId) incrementLateNightTripsCountBy
+    Nothing -> pure ()
+  where
+    isLateNightRide :: (EsqDBFlow m r, EsqDBReplicaFlow m r) => DR.Ride -> m Int
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs:118]: can use boolean simply

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore/Types.hs [hunk 4]
```diff
@@ -49,6 +52,12 @@ data DriverRideRequeset
       }
   | OnDriverCancellation
       { merchantId :: Id DM.Merchant,
-        driverId :: Id DP.Person
+        driverId :: Id DP.Person,
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore/Types.hs:55]: Can you please rename this - `DriverRideRequeset`

File: Backend/dev/migrations/dynamic-offer-driver-app/0209-added-stats-column-and-driver-hometown.sql [hunk 5]
```diff
@@ -0,0 +1,9 @@
+ALTER TABLE atlas_driver_offer_bpp.driver_stats ADD COLUMN total_earnings int DEFAULT 0;
+ALTER TABLE atlas_driver_offer_bpp.driver_stats ADD COLUMN bonus_earned int DEFAULT 0;
+ALTER TABLE atlas_driver_offer_bpp.driver_stats ADD COLUMN late_night_trips int DEFAULT 0;
+ALTER TABLE atlas_driver_offer_bpp.driver_stats ADD COLUMN earnings_missed int DEFAULT 0;
+
+ALTER TABLE atlas_driver_offer_bpp.person ADD COLUMN hometown character varying(255);
+ALTER TABLE atlas_driver_offer_bpp.person ADD COLUMN languages_spoken text[] DEFAULT '{}'::text[];
```
Comment (0utkarsh) [inline Backend/dev/migrations/dynamic-offer-driver-app/0209-added-stats-column-and-driver-hometown.sql:7]: This isn't json, right?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Vehicle.hs [hunk 6]
```diff
@@ -70,6 +71,15 @@ updateVehicleRec vehicle = do
       ]
     where_ $ tbl ^. VehicleTId ==. val (toKey vehicle.driverId)
 
+updateVehicleName :: Maybe Text -> Id Person -> SqlDB ()
+updateVehicleName vehicleName driverId = do
+  Esq.update $ \tbl -> do
+    set
+      tbl
+      [ VehicleVehicleName =. val vehicleName
+      ]
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Vehicle.hs:80]: What is `VehicleName`? Also, update `updatedAt` column

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/FareParameters.hs [hunk 7]
```diff
@@ -44,3 +45,38 @@ findAllIn fareParamIds =
...
+findDriverSelectedFareEarnings :: Transactionable m => [Id FareParameters] -> m Money
+findDriverSelectedFareEarnings fareParamIds =
+  mkSum
+    <$> Esq.findAll do
+      fareParamFile <- from $ table @FareParametersT
+      where_ $ fareParamFile ^. FareParametersId `in_` valList (map getId fareParamIds)
+      pure (sum_ $ fareParamFile ^. FareParametersDriverSelectedFare :: SqlExpr (Esq.Value (Maybe Money)))
+  where
+    mkSum [Just value] = value
+    mkSum _ = 0
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/FareParameters.hs:71]: repetition

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverStats.hs [hunk 8]
```diff
@@ -104,3 +120,22 @@ getDriversSortedOrder mbLimitVal =
...
+        DriverStatsBonusEarned =. (tbl ^. DriverStatsBonusEarned) +. val increasedBonus
+      ]
+    where_ $ tbl ^. DriverStatsDriverId ==. val (toKey $ cast driverId)
+
+incrementLateNightTrips :: Id Driver -> Int -> SqlDB ()
+incrementLateNightTrips driverId tripCount = do
+  Esq.update $ \tbl -> do
+    set
+      tbl
+      [ DriverStatsLateNightTrips =. (tbl ^. DriverStatsLateNightTrips) +. val tripCount
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverStats.hs:139]: can you add updated timestamp everywhere?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs [hunk 9]
```diff
@@ -122,3 +164,30 @@ incrementOrSetTotaRides driverId driverStats = do
...
   pure $ driverStats {DS.totalRidesAssigned = Just $ fromMaybe 0 driverStats.totalRidesAssigned + incrementTotaRidesBy}
+
+getDriverStats :: (EsqDBFlow m r, EsqDBReplicaFlow m r) => Maybe DS.DriverStats -> Id DP.Person -> Maybe Money -> m DS.DriverStats
+getDriverStats Nothing driverId _ = createDriverStat driverId
+getDriverStats (Just driverStats) driverId rideFare = do
+  updatedTotalRideCount <- getTotalRideCount
+  cancelledCount <-
+    case driverStats.ridesCancelled of
+      Nothing -> Esq.runInReplica $ BCRQ.findAllCancelledByDriverId driverId
+      Just cancelledCount -> pure $ cancelledCount + 1
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs:175]: why +1?

**Review by 0utkarsh — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/DriverInformation.hs (hunk 1 above)
Comment (prashant601) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/DriverInformation.hs:90]: In driverStats , cancelled count was maybe and I am just using that count

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs (hunk 9 above)
Comment (prashant601) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs:175]: As we discussed, This +1 is for incrementing the cancelled ride count, and this is implemented by Piyush, I just refactored this part.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Vehicle.hs (hunk 6 above)
Comment (prashant601) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Vehicle.hs:80]: As discussed already , this vehicle name is the name for the vehicle given by the driver, use case was to improve the ux

File: Backend/dev/migrations/dynamic-offer-driver-app/0209-added-stats-column-and-driver-hometown.sql (hunk 5 above)
Comment (prashant601) [reply Backend/dev/migrations/dynamic-offer-driver-app/0209-added-stats-column-and-driver-hometown.sql:7]: it's creating an empty list as the default value

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs [hunk 10]
```diff
@@ -91,18 +81,68 @@ eventPayloadHandler DST.OnDriverCancellation {..} = do
...
+              pure $ fromMaybe 0 (booking.fareParams.driverSelectedFare) + fromMaybe 0 (booking.fareParams.customerExtraFee) + 10 -- 10 is the pickup charge
+            Nothing -> pure 0
+    incrementLateNightTripsCountBy <-
+      if isNotBackFilled driverStats
+        then do
+          Esq.runInReplica $ FPQ.findAllLateNightRides farePramIds
+        else isLateNightRide ride
+
+    Esq.runNoTransaction $ do
+      DSQ.incrementTotalEarningsAndBonusEarned (cast driverId) incrementTotaEarningsBy (incrementBonusEarningsBy + Money (length farePramIds * 10))
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs:112]: these can be done in one query

**Review by piyushKumar-1 — APPROVED** [review]

**Pull request merged**